### PR TITLE
Page authenticate role

### PR DESCRIPTION
**Pull request by siddhsuresh**
Closes: #2399

### What are the changes and their implications?

- [x] add option `role` to the `authenicate` property to authorize users depending on their role

Values that can be passed to `role` property is `string` or an `array` of `strings`

Used the following convention [mentioned in the issue]
```jsx
// user must be logged in
Page.authenticate = true 
// user must be logged in and redirects otherwise
Page.authenticate = { redirectTo: "/login" }
// user must be logged in and with role "ADMIN" or "USER"
Page.authenticate = { role: ["ADMIN", "USER"] }
// user must be logged in with role "ADMIN" and redirects otherwise
Page.authenticate = { role: "ADMIN", redirectTo: "/login" }
```

## Feature Checklist

- [x] Changeset added (run `pnpm changeset` in the root directory)
- [x] fixed test in  `packages\blitz-auth\src\client\auth-client.test.ts` to show `role` as public
- [x] Documentation added/updated (submit PR to [blitzjs.com repo `main` branch](https://github.com/blitz-js/blitzjs.com))

### CODE REVIEW

File: packages/blitz-auth/src/client/index.tsx [hunk 1]
```diff
@@ -268,7 +290,10 @@ function withBlitzAuthPlugin<TProps = any>(Page: ComponentType<TProps> | BlitzPa
 
     let {authenticate, redirectAuthenticatedTo} = getAuthValues(Page, props)
 
-    useAuthorizeIf(authenticate === true)
+    useAuthorizeIf(
+      !authenticate ? false : typeof authenticate === "boolean" ? authenticate === true : true,
```
Comment (siddhsuresh) [inline packages/blitz-auth/src/client/index.tsx:294]:
```suggestion
      useAuthorizeIf(authenticate === true),
```

File: packages/blitz-auth/src/client/index.tsx [hunk 2]
```diff
@@ -268,7 +290,10 @@ function withBlitzAuthPlugin<TProps = any>(Page: ComponentType<TProps> | BlitzPa
 
     let {authenticate, redirectAuthenticatedTo} = getAuthValues(Page, props)
 
-    useAuthorizeIf(authenticate === true)
+    useAuthorizeIf(
+      useAuthorizeIf(authenticate === true),
```
Comment (siddhsuresh) [inline packages/blitz-auth/src/client/index.tsx:294]:
```suggestion
      authenticate === true,
```

**Review by dillonraphael — APPROVED** [review]

**Comment by dillonraphael** [comment]
@siddhsuresh Were you able to update the docs with this? Great job as always!

**Comment by siddhsuresh** [comment]
> @siddhsuresh Were you able to update the docs with this? Great job as always!

@dillonraphael I linked the updated docs PR

File: apps/web/pages/page-with-redirect.tsx [hunk 3]
```diff
@@ -13,6 +13,6 @@ const PageWithRedirect = () => {
   )
 }
 
-PageWithRedirect.authenticate = {redirectTo: "/"}
+PageWithRedirect.authenticate = {role: ["ADMIN"], redirectTo: "/"}
```
Comment (siddhsuresh) [inline apps/web/pages/page-with-redirect.tsx:16]:
```suggestion
PageWithRedirect.authenticate = {redirectTo: "/"}
```

File: apps/web/app/mutations/createUser.ts [hunk 4]
```diff
@@ -6,7 +6,6 @@ export default async function createUser(
   ctx: Ctx,
 ): Promise<User> {
   ctx.session.$authorize()
```
Comment (siddhsuresh) [inline apps/web/app/mutations/createUser.ts:8]:
```suggestion
  ctx.session.$authorize()
  
```

**Comment by dillonraphael** [comment]
Sorry @siddhsuresh, can we just get the conflicts resolved? Going to get this merged and released.

**Comment by siddhsuresh** [comment]
> Sorry @siddhsuresh, can we just get the conflicts resolved? Going to get this merged and released.

@dillonraphael I resolved the merge conflict

**Pull request merged**